### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by Mayank447**
Implemented different kernels and GP Models for bond length, angles and dihedrals.

### CODE REVIEW

File: boptimol/GP_MODEL.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+from botorch.fit import fit_gpytorch_mll
+from botorch.acquisition import qKnowledgeGradient
+from gpytorch.mlls import ExactMarginalLogLikelihood
+from gpytorch import kernels as gpykernels
+from gpytorch import means as gpymeans
+
+from boptimol.molecule import Molecule
+from boptimol.mean_potentials import MorseMean, QuadraticMean
+
+class Bond_Lengths(gpytorch.models.ExactGP):
```
Comment (ghutchis) [inline boptimol/GP_MODEL.py:19]: I'd like to see a lot more docstrings so that others can understand what the classes do, etc.

File: boptimol/GP_MODEL.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+    Uses BOTorch to pick the next sample using Expected Improvement
+
+    Args:
+        observed_X: Observed coordinates
+        observed_y: Observed energies
+    Returns:
+        Next coordinates to try
+    """
+    
+    likelihood = gpytorch.likelihoods.MultitaskGaussianLikelihood(num_tasks=1).to(device=device)
```
Comment (ghutchis) [inline boptimol/GP_MODEL.py:48]: Does this need to be Multitask if we're always setting `num_tasks` to 1?

File: boptimol/GP_MODEL.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    gp = Bond_Lengths(observed_X, observed_y, likelihood)
+    mll = gpytorch.mlls.ExactMarginalLogLikelihood(gp.likelihood, gp)
+    fit_gpytorch_mll(mll)
+    
+    qKG = qKnowledgeGradient(gp, num_fantasies=4)
```
Comment (ghutchis) [inline boptimol/GP_MODEL.py:53]: This seems like you're experimenting with KnowledgeGradient, which is great. But then why have the logEI below that?

I think we need to have a way to pass or set the acquisition function for all these classes - so that we can easily benchmark the different acquisition methods.

File: boptimol/GP_MODEL.py (hunk 2 above)
Comment (ghutchis) [reply boptimol/GP_MODEL.py:48]: If it's because we're modeling energy + dihedral, please explain that in the comments

File: boptimol/GP_MODEL.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+    """
+    
+    mean_module = gpymeans.ConstantMean(batch_shape=torch.Size())
+    covar_module = gpykernels.ScaleKernel(gpykernels.RBFKernel())
+    
+    gp = SingleTaskGP(observed_X, observed_y, covar_module=covar_module, mean_module=mean_module)
+    mll = ExactMarginalLogLikelihood(gp.likelihood, gp).to(device=device)
+    fit_gpytorch_mll(mll)
+    
+    aq = LogProbabilityOfImprovement(gp, best_f=torch.max(observed_y))
```
Comment (ghutchis) [inline boptimol/GP_MODEL.py:90]: Again, it would be great to set the acquisition function systematically.

File: boptimol/GP_MODEL.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+    Uses BOTorch to pick the next sample using Expected Improvement
+
+    Args:
+        observed_X: Observed coordinates
+        observed_y: Observed energies
+    Returns:
+        Next coordinates to try
+    """
+    
+    mean_module = gpymeans.ConstantMean(batch_shape=torch.Size())
```
Comment (ghutchis) [inline boptimol/GP_MODEL.py:117]: Is this better than a quadratic mean?

File: boptimol/GP_MODEL.py [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+    gp = SingleTaskGP(observed_X, observed_y, covar_module=covar_module, mean_module=mean_module)
+    mll = ExactMarginalLogLikelihood(gp.likelihood, gp).to(device=device)
+    fit_gpytorch_mll(mll)
+    
+    aq = LogProbabilityOfImprovement(gp, best_f=torch.max(observed_y))
+    candidate, acq_value = optimize_acqf(
+        aq, 
+        bounds=bounds,
+        q=1, 
+        num_restarts=64,
```
Comment (ghutchis) [inline boptimol/GP_MODEL.py:129]: NUM_RESTARTS

File: boptimol/GP_MODEL.py [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+    candidate, acq_value = optimize_acqf(
+        aq, 
+        bounds=bounds,
+        q=1, 
+        num_restarts=64, 
+        raw_samples=64
+    )
+    return candidate.detach().numpy()[0, :]
+
+
```
Comment (ghutchis) [inline boptimol/GP_MODEL.py:134]: Is there a bond angle with derivatives?

File: boptimol/GP_MODEL.py [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+    mean_module = gpymeans.ConstantMean(batch_shape=torch.Size())
+    covar_module = gpykernels.ScaleKernel(gpykernels.ProductStructureKernel(
+                num_dims = observed_X.shape[1],
+                base_kernel=gpykernels.PeriodicKernel()))
+    
+    gp = SingleTaskGP(observed_X, observed_y, covar_module=covar_module, mean_module=mean_module)
+    mll = ExactMarginalLogLikelihood(gp.likelihood, gp).to(device=device)
+    fit_gpytorch_mll(mll)
+    
+    aq = LogExpectedImprovement(gp, best_f=torch.max(observed_y))
```
Comment (ghutchis) [inline boptimol/GP_MODEL.py:161]: Again, you have some logPI and logEI and other acquisition functions all mixed together.

File: boptimol/GP_MODEL.py [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+    candidate, acq_value = optimize_acqf(
+        aq, 
+        bounds=bounds,
+        q=1, 
+        num_restarts=64,
```
Comment (ghutchis) [inline boptimol/GP_MODEL.py:166]: NUM_RESTARTS

File: boptimol/GP_MODEL.py [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+    mll = ExactMarginalLogLikelihood(gp.likelihood, gp).to(device=device)
+    fit_gpytorch_mll(mll)
+    
+    aq = LogExpectedImprovement(gp, best_f=torch.max(observed_y))
+    candidate, acq_value = optimize_acqf(
+        aq, 
+        bounds=bounds,
+        q=1, 
+        num_restarts=64, 
+        raw_samples=64
```
Comment (ghutchis) [inline boptimol/GP_MODEL.py:167]: RAW_SAMPLES

File: boptimol/GP_MODEL.py [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+
+    Args:
+        observed_X: Observed coordinates
+        observed_y: Observed energies
+    Returns:
+        Next coordinates to try
+    """
+
+    # Clipping the energies if needed
+    observed_y = np.clip(observed_y, -np.inf, 2 + np.log10(np.clip(observed_y, 1, np.inf)))
```
Comment (ghutchis) [inline boptimol/GP_MODEL.py:190]: I'm not 100% certain this is a good idea anymore. We should definitely test if it's a benefit.

File: boptimol/GP_MODEL.py [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+    candidate_bond_angles = select_next_point_bond_angles(bounds[:, mol.end_bonds: mol.end_angles],
+                             train_X[:, mol.end_bonds: mol.end_angles], train_Y, device)
+    
+    candidate_dihedrals = select_next_point_dihedrals(bounds[:, mol.end_angles: mol.degrees_of_freedom],
+                             train_X[:, mol.end_angles: mol.degrees_of_freedom], train_Y, device)
+    
+    candidate = np.concatenate((candidate_bond_lengths, candidate_bond_angles, candidate_dihedrals))
+    
+    
+    # Solve the optimization problem
```
Comment (ghutchis) [inline boptimol/GP_MODEL.py:215]: I think we can drop all this, correct?

File: boptimol/GP_MODEL.py [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+    # Convert inputs to torch arrays
+    train_X = torch.tensor(observed_X, dtype=torch.float64, device=device)
+    train_y = torch.tensor(observed_y, dtype=torch.float64, device=device)
+    
+    # Making this a maximization problem
+    train_Y = standardize(-1* train_y[:, None])
+
+    candidate_bond_lengths = select_next_point_bond_lengths(bounds[:, : mol.end_bonds],
```
Comment (ghutchis) [inline boptimol/GP_MODEL.py:203]: This definitely needs some comments - or at least assign the various train_X and bounds slices to named variables, e.g.
```
bond_bounds = bounds[:, : mol.end_bonds]
bond_X = train_X[:, : mol.end_bonds]
# …
angle_bounds = bounds[:, mol.end_bonds: mol.end_angles]
# …
```

It's important to think who will read the code .. which includes yourself in a few weeks.

File: boptimol/initializer.py [hunk 14]
```diff
@@ -0,0 +1,38 @@
+from typing import List, Optional
+import logging
+
+import numpy as np
+from boptimol import Molecule
+
+logger = logging.getLogger(__name__)
+
+# Making initial guesses for the molecule (Active Learning)
```
Comment (ghutchis) [inline boptimol/initializer.py:9]: This is a good idea, since obviously we'll expand this.

File: boptimol/initializer.py [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+def initial_guess(mol: Molecule, init_steps: int, out_dir: str) -> np.ndarray:
+    
+    ''' Return initial coordinates guess based on the molecule (Acti)
+    Args:
+        mol: A Molecule Class Object
+        init_step: No of initial guesses to be returned
+    Returns:
+        Initial Guess Coordinates to be tried
```
Comment (ghutchis) [inline boptimol/initializer.py:17]: initial guess .. as z-matrix, correct? Probably better to say that.

File: boptimol/initializer.py [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+    '''
+    
+    start_coords = mol.parameters
+    start_energy = mol.energy(start_coords)
+    
+    init_guesses = np.random.normal(start_coords, 0.1, size=(init_steps, len(start_coords))) #TODO
```
Comment (ghutchis) [inline boptimol/initializer.py:23]: This might be good to separate bonds and angles and dihedrals already. I know I wrote this, but +/- 0.1 Å for a bond isn't a big deal. 0.1 radian is ~6° which is fairly large for angles.. maybe 0.05 for angles and dihedrals?

File: boptimol/mean_potentials.py [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+        res = self.depth * (1 - res)**2
+        if self.bias is not None:
+            res = res + self.bias
+        res.view(-1)
+        return res
+
+    
+class QuadraticMean(Mean):
+    
+    """ A configurable mean function modelled as a Quadratic
```
Comment (ghutchis) [inline boptimol/mean_potentials.py:39]: Might be worth indicating the formula in the docs.

File: boptimol/mean_potentials.py [hunk 18]
```diff
@@ -0,0 +1,64 @@
+from gpytorch.means import ConstantMean, Mean
+import torch
+
+class MorseMean(Mean):
+    
+    """ A configurable mean function using the Morse potential
+
+    Uses BOTorch to pick the next sample using Expected Improvement
```
Comment (ghutchis) [inline boptimol/mean_potentials.py:8]: Might be worth indicating the formula in the docs, since not everyone knows this.

File: boptimol/mean_potentials.py [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+            res = res + self.bias
+        res.view(-1)
+        return res
+
+    
+class QuadraticMean(Mean):
+    
+    """ A configurable mean function modelled as a Quadratic
+
+    Uses BOTorch to pick the next sample using Expected Improvement
```
Comment (ghutchis) [inline boptimol/mean_potentials.py:41]: using the acquisition function (since it's not always EI).

File: boptimol/mean_potentials.py [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+
+    Args:
+        batch_shape: train data
+        bias: an energy offset term
```
Comment (ghutchis) [inline boptimol/mean_potentials.py:45]: center

File: boptimol/molecule.py [hunk 21]
```diff
@@ -75,24 +81,38 @@ def __init__(self, filename: str, calculator: Calculator = None):
...
-        lower_bounds = np.zeros(degrees_of_freedom)
-        upper_bounds = np.zeros(degrees_of_freedom)
+        self.degrees_of_freedom = self.end_angles + len(self.dihedrals)
 
+        lower_bounds = np.zeros(self.degrees_of_freedom)
+        upper_bounds = np.zeros(self.degrees_of_freedom)
+        
+        '''chemcoord helper functions'''
+        #getElements in a bond
+        getElement = lambda idx: (self.zmat['atom'][idx],
```
Comment (ghutchis) [inline boptimol/molecule.py:91]: Do you want to make this a full method (e.g., getBondElements() returns a tuple?) .. it seems like it would be useful elsewhere.

File: boptimol/optimizer.py [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+from boptimol.GP_MODEL import select_next_points_botorch
+from boptimol.initializer import initial_guess
+
+logger = logging.getLogger(__name__)
+
+
+def run_optimization(mol: Molecule, n_steps: int, 
+                     init_steps: int, out_dir: Optional[Path]) -> np.ndarray:
+    
+    """Optimize the structure of a molecule by iteratively changing the dihedral angles
```
Comment (ghutchis) [inline boptimol/optimizer.py:23]: Iteratively changing the z-matrix parameters. (I think I copied this from the conformer optimizer which only optimizes dihedrals.)

File: boptimol/optimizer.py [hunk 23]
```diff
@@ -0,0 +1,114 @@
...
+        # Update the search space
+        observed_coords = np.vstack([observed_coords, next_coords])
+        observed_energies.append(energy - start_energy)
+        best_energy = np.min(observed_energies)
+
+        # Check for convergence
+        # TODO: check for RMSD of the parameters
+        #if step>0 and cc.xyz_functions.isclose(observed_energies[step],observed_energies[step-1]):
+            
+        if np.abs(energy - best_energy) < 1e-3:
```
Comment (ghutchis) [inline boptimol/optimizer.py:100]: Is there a good way to get the derivative / gradient as well?

File: environment.yml [hunk 24]
```diff
@@ -2,13 +2,12 @@ name: confopt
 channels:
   - defaults
   - conda-forge
-  - psi4
+  - mcocdawc
```
Comment (ghutchis) [inline environment.yml]: If we're using `pip` to install `chemcoord` we don't need this channel. He also said he'll delete it once the conda-forge package is merged.

File: test.ipynb
Comment (ghutchis) [inline test.ipynb:1]: I'm not sure we want to merge this?

Ideally, we'd add some `pytest` tests for all of the molecules and expected energies as another step.

**Review by ghutchis — CHANGES_REQUESTED** [review]

File: boptimol/molecule.py (hunk 21 above)
Comment (Mayank447) [reply boptimol/molecule.py:91]: These functions were much used right now, so thats why a created a lambda function. I will actually make these changes in chemcoord and make a pull request.

File: test.ipynb
Comment (Mayank447) [reply test.ipynb:1]: Sorry, by mistake I stagged this file as well.

File: environment.yml (hunk 24 above)
Comment (Mayank447) [reply environment.yml]: I fixed it.

File: boptimol/molecule.py (hunk 21 above)
Comment (ghutchis) [reply boptimol/molecule.py:91]: In the meantime, as I mentioned, I'd put this into the `Molecule` class - you will definitely use it in the initial guess code.
